feat!: GST 2026 changes

hairmare opened the pull request:
This PR contains the changes for GST 2026 as well as the configuration for doing prereleases.

Once merged, we will be moving towards a stable v1 version of the tool.

* replaces #878
* replaces #879
* replaces #885
* fixes #825
* fixes #863
* fixes #528
* fixes #75 
* fixes #36
* related https://github.com/radiorabe/rabe-zabbix/pull/166

hairmare commented on suisa_sendemeldung/suisa_sendemeldung.py:
These need to be updated for GST 2026, there is also an additional lime green colour that is used for fields needed in subsidiary purposes when there is no ISRC for a record.

hairmare commented on suisa_sendemeldung/suisa_sendemeldung.py:
```diff
@@ -794,17 +555,13 @@ def get_email_attachment(filename: str, filetype: str, data: Any) -> MIMEBase:
-    elif filetype == "csv":
+    maintype = "application"
+    subtype = "vnd.ms-excel"
+    payload: str | bytes = str(data)
```
We need to keep using `getvalue()`:
```suggestion
    payload: str | bytes = data.getvalue()
```

hairmare commented:
The v1.0.0-alpha.5 is currently configured to send additional test mails from our porduction infrastructure. We will be getting a test mail for Dec 2025 as well as Jan 2026, giving us enough time to fix anything if the need arises.

Once the Dec 2025 mail has been received by SUSIA (in mid Jan), we'll proceed with merging this,

paraenggu commented on suisa_sendemeldung/settings.py:
```diff
@@ -0,0 +1,222 @@
+- ISRC der benützten Aufnahme (sofern zusammen mit der Aufnahme vom Lieferanten der Aufnahme irgendeiner Form mitgeteilt bzw. mitgeliefert)
+- Label
+
+Die folgenden subsidiären Angaben melden wir in allen Fällen wo diese vom Lieferanten der Aufnahme mitgeteilt bzw. mitgeliefert wurden, insbesondere in Fällen wo der Lieferant der Aufnahme uns keine ISRC mitgeteilt bzw. mitgeliefert hat.
```
```suggestion
Die folgenden subsidiären Angaben werden in allen Fällen gemeldet, in denen diese vom Lieferanten der Aufnahme mitgeteilt oder mitgeliefert wurden, insbesondere in Fällen, in denen der Lieferant der Aufnahme keine ISRC mitgeteilt oder mitgeliefert hat.
```

Using passive voice as within the rest of the mail text.

hairmare replied:
The passus "mitgeteilt bzw. mitgeliefert" was taken straight from the tarif, but changing it in our mail makes sense.

paraenggu commented on suisa_sendemeldung/suisa_sendemeldung.py:
```diff
@@ -553,46 +293,42 @@ def get_isrc(music: Any) -> str:  # noqa: ANN401
-        "Label Code",
-        "EAN/GTIN",
-        "Identifikationsnummer",
+        "Veröffentichungsland",
```
```suggestion
        "Veröffentlichungsland",
```

Spelling correction.

hairmare replied:
Same here, the template that was ratified by ESchK introduces this error.

paraenggu replied:
Indeed, that term is misspelled in the published template as well. IMO it should be corrected, as it is clearly incorrect. If there are any complaints, the typo can be reintroduced, though I highly doubt that will occur. 😉

hairmare commented on suisa_sendemeldung/suisa_sendemeldung.py:
```diff
@@ -553,46 +293,42 @@ def get_isrc(music: Any) -> str:  # noqa: ANN401
-        "Track Nummer",
-        "Genre",
-        "Programm",
+        "Erstveröffentichungsdatum",
```
This follows the template to the letter!

paraenggu replied:
> This follows the template to the letter!

it's correctly spelled as "Erstveröffentlichungsdatum" in the template (_Vorlage Sendelisten Radio_) at the end of the [Gemeinsamer Tarif S 2026 – 2028](https://www.suisa.ch/dam/jcr:ea035e73-ce1a-48b7-9269-3f6aedc4057f/GTS_2026-2028_GER.pdf) document, This must have been introduced from somewhere else.

<img width="482" height="355" alt="image" src="https://github.com/user-attachments/assets/d6f7518f-dde3-435e-9a48-f789cf2dd457" />

Regardless of the source, I don't think that it makes sense to introduce typos here.

hairmare commented on suisa_sendemeldung/suisa_sendemeldung.py:
```diff
@@ -603,7 +339,7 @@ def get_csv(data: dict, args: ArgparseNamespace) -> str:
         timestamp = datetime.strptime(metadata.get("timestamp_local"), ACRClient.TS_FMT)  # noqa: DTZ007
 
-        ts_date = timestamp.strftime("%Y%m%d")
+        ts_date = timestamp.strftime("%Y-%m-%d")
```
Exactly, the OOXML uses a proper ISO date under the hood and has a format specifier to render it as DD.MM.YYYY on display.

hairmare commented on suisa_sendemeldung/suisa_sendemeldung.py:
```diff
@@ -719,26 +452,40 @@ def get_xlsx(data: Any, args: ArgparseNamespace) -> BytesIO:  # noqa: ANN401
     border = Border(top=side, left=side, right=side, bottom=side)
-    fill = PatternFill("solid", bgColor="d9d9d9", fgColor="d9d9d9")
+    required_fill = PatternFill("solid", bgColor="d9d9d9", fgColor="d9d9d9")
+    subsdiary_fill = PatternFill("solid", bgColor="ebf1de", fgColor="ebf1de")
```
i checked the inofficial original for that actual color code (instead of using the eyedropper on the PDF):
```suggestion
    subsdiary_fill = PatternFill("solid", bgColor="aef0dd", fgColor="aef0dd")
```

paraenggu replied:
> i checked the inofficial original for that actual color code (instead of using the eyedropper on the PDF):

What is the "unofficial original"?

I compared and took the colors from the officially published [Gemeinsamer Tarif S 2026 – 2028](https://www.suisa.ch/dam/jcr:ea035e73-ce1a-48b7-9269-3f6aedc4057f/GTS_2026-2028_GER.pdf) document. There it's `bfbfbf` for the mandatory columns (_grey_, "zwingend") and `ebf1de` for the mandatory if no ISRC columns (_green_, zwingend, falls kein ISRC):

<img width="817" height="243" alt="image" src="https://github.com/user-attachments/assets/c28da166-bb93-45b9-9e5f-36ac82cd57ba" />

hairmare commented on suisa_sendemeldung/suisa_sendemeldung.py:
```diff
@@ -644,7 +380,7 @@ def get_csv(data: dict, args: ArgparseNamespace) -> str:
         upc = music.get("external_ids", {}).get("upc", "")
         release_date = funge_release_date(music.get("release_date", ""))
```
This is "Erstveröffentichungsdatum" and it's being adjusted by `reformat_start_date_in_xlsx`.

There should be a bunch of examples for it in the upcoming test mail.

hairmare commented:
@paraenggu I merged most changes and added some feedback for others. I updated prod to use v1.0.0-alpha.6 in the next test run.

wrt "Erstveröffentichungsdatum" i think we get to decide if we want to keep the typo that was most likely proposed by SUISA and ratified by ESchK. It's kinda disturbing how many people were involved in the new GST and none of them seems to have had a working spellchecker.

Anyway, i'm not sure if keeping the typo might be to petty, so we could consider fixing the spelling. In this context we might also refer to the text of the GST, where the column is called "(Erst-)Veröffentlichungsdatum bzw. -jahr".

My understanding is that this text won't have any impact on SUISA's ability to process our reports, hence i would propbably make it match the template, leaving the typo in the report (as currently implemented).

paraenggu commented:
> I merged most changes and added some feedback for others. I updated prod to use v1.0.0-alpha.6 in the next test run.

Great, thank you!


> wrt "Erstveröffentichungsdatum" i think we get to decide if we want to keep the typo that was most likely proposed by SUISA and ratified by ESchK.

That one is correctly spelled  in the template (see [comment r2650351762](#discussion_r2650351762)), but "Veröffentlichungsland" is still wrong in the template.

As mentioned, I would prefer that the corrected spelling be used from the beginning, rather than including typos in the newly generated output. If any complaints arise, the typo can still be reintroduced as needed.

spameier commented on suisa_sendemeldung/suisa_sendemeldung.py:
```diff
@@ -19,328 +18,69 @@
     """
     msgs = []
-    # check length of bearer_token
-    if not len(args.bearer_token) >= _ACRTOKEN_MAXLEN:
```
We loose some validation of user provided flags, imo this is fine for our automation use case. I would still prefer to keep the logic so user errors are caught before calling the API.

hairmare commented:
Feedback from mail (sent by @paraenggu)
> # Mail-Text
>
> Abstimmung:
>
>> [...] sowie der Abstimmung zwischen SUISA und TEST Radio Bern RaBe TEST.
>
>Würde diesen Teil streichen, es gab keine Abstimmung wie beim letzten
>mal als wir den automatischen Report einführten.

- [ ] 👍 

> ## Subsidiäre Angaben:
>
> > Die folgenden subsidiären Angaben werden in allen Fällen gemeldet, in denen diese vom Lieferanten der Aufnahme mitgeteilt oder mitgeliefert wurden, insbesondere in Fällen, in denen der Lieferant der Aufnahme keine ISRC mitgeteilt oder mitgeliefert hat.
>
> Ich schlage vor den Text hier etwas anzupassen. Wir liefern die Angaben
> ja immer sofern verfügbar und nicht nur, oder wie geschrieben
> insbesondere, bei fehlender ISRC. Ebenfalls empfehle ich hier die
> Formulierung bezüglich "verfügbar/zugänglich" aus dem Tarif aufzugreifen.
> 
> Im Tarif steht zu den subsidiären Angaben folgendes:
>
> > Falls der ISRC nicht gemeldet wird, sind subsidiär folgende Angaben zu machen, sofern diese beim Sender verfügbar oder für ihn leicht zugänglich sind:
>
> Vorschlag:
>
> Ebenfalls sind folgende subsidiäre Angaben enthalten, sofern sie für den
> Sender verfügbar oder leicht zugänglich sind:
>
> - EAN / GTIN (European Article Number / Global Trade Item Number)
> - Albumtitel / Titel des Tonträgers
> - [...]

- [ ] The ideas was indeed that the whole "mitgeteilt oder mitgeliefert" would also cover this, but using "verfügbar oder leicht zugänglich" makes more sense.

> ## Zeilenlänge (kosmetisch):
> Die Zeilenlänge der Mail war zuvor auf eine maximale Länge
> vereinheitlicht. Nun hat es gewisse Abschnitte welche länger sind.

- [ ] I'll probably change it to be a single line per paragraph, thus letting the MUA do the formatting.

> # Sendemeldung
>
> ## Dateinahme:
>
> Der Dateiname heisst test_2025-12-02.xlsx statt test_2025-12.xlsx - Ist
> dies bedingt durch dein Test oder hat sich da noch ein "DD" (Tag) im
> Dateinamen eingeschlichen?

- [ ] The "last month" mode seems to be broken somewhere, this affects both the filename as well as the start and end date.

> ## Spaltenbezeichnung:
>
> Die Splate "L" ist mit "EAN/GTIN" bezeichnet, in der Vorlage heisst
> diese "EAN / GTIN" (Leerzeichen vor und nach dem Schrägstrich).

- [ ] 👍 

> ## Farbe:
>
> Farbe für "zwingend, falls kein ISRC" Felder stimmt nicht mit dem "grün"
> aus der öffentlich publizierten Vorlage überein, müsste #ebf1de sein.

- [ ] 👍 

> ## Start- & Endzeitpunkt sind falsch:
>
> Der Rapport beginnt am 02.12.2025 statt am 01.12. und gehen bis zum
> 01.01.2026 08:53:26 statt bis zum 31.12 23:59

- [ ] Seems to be related to the filename issue above, I'll double check this and try to figure out where i missed a test.

> ## Sendedatum:
>
> In der Spalte Sendedatum ist neu ebenfalls eine Uhrzeit enthalten, z.B.:
> "02.12.2025 00:00:31"
> Hier sollte nur das Datum vorhanden sein, damit es nicht zu Problemen
> auf Seite SUISA-Import kommt.

- [ ] Needs fixing, I'm not sure where i introduced it yet.

> ## Leere "zwingend, falls kein ISRC" Felder:
>
> Die folgenden Felder, welche als "zwingend falls kein ISRC und
> verfügbar" eingestuft sind, sind immer leer:
>
> - Aufnahmedatum
> - Aufnahmeland
> - Katalog-Nummer / CD ID
> - Werkverzeichnisangaben
> - Bestellnummer
>
> Gibt es hierzu von Seite ACRCloud weiterhin keine zusätzlichen Daten?

- [ ] I'll have to check our documentation to see if we have more info on this.

> ## Identifikationsnummer:
>
> Der Sendenamen innerhalb der crid scheint mittlerweile der korrekten
> Uhrzeit zu entsprechen - hurra 🥳

🎉 

I'll fix these and then resend the Dec mail!

hairmare commented:
New fields:

> Aufnahmedatum

We only have "release_date" in our records, information on the date of the recording does not seem to be publicaly available.

> Aufnahmeland

There are some complex data structures under "release_by_territories", they don't seem to be relevant to the country of recording but rather contain deal related information from the music industry at large. The data structure in our results is also not consistent and we would require better documentation to make sense of the data.

> Katalog-Nummer / CD ID

The CD_ID is available in some album records, so we can add it.

The tarif describes the "Katalog-Nummer" as "Katalognummer des Tonträgers beim Label". The way our provisioning of samples (Bemusterung) works, it doesn't look like we have this info as it seems to be specific to individual labels.

> Werkverzeichnisangaben

This would be something like "Opus n", it's rarely applicable outside of classical music and also doesn't seem to be available in ACR.

> Bestellnummer

I have zero clues what this is, the tarif doesn't seem to have any further details.

hairmare commented:
last-month mode seems to have not been addressed in the beta image

* [ ] fix filename issue again (needs to be rabe_<year>-<month> not rabe_<year>-<month>-<day>
* [ ] fix start date issue again (needs to start at the 1st of the last month, no at the day of execution)

both cases are related to the last-month mode not properly doing it's thing, i did some fixing in this area for beta.1, but my fixes seem to have missed the mark
